tests/utils/test_neighbors.py

```diff
@@ -0,0 +1,339 @@
+            source_site = structure.sites[src_idx]
+            for neighbor_site in neighbor_sites:
+                # Make sure the neighbors are no more than one shell away.
+                assert np.all(np.abs(np.array(neighbor_site.image)) <= 1.), \
+                    "A neighbor site is out of range. Review test code!"
+
+                dst_idx = neighbor_site.index
+                displacement = neighbor_site.coords - source_site.coords
+
+                shift = np.dot(np.array(neighbor_site.image), structure.lattice.matrix)
+
+                neighbors = Neighbors(source_index=src_idx,
+                                      destination_index=dst_idx,
+                                      displacement=displacement,
+                                      shift=shift)
+                list_neighbors.append(neighbors)
+
+        structure_neighbors.append(list_neighbors)
+
+    return structure_neighbors
+
+
+@pytest.fixture
+def expected_neighbor_indices_and_displacements(expected_neighbors):
+
+    list_source_indices = []
+    list_dest_indices = []
+    list_displacements = []
+    list_shifts = []
+    for list_neighbors in expected_neighbors:
+        source_indices = torch.tensor([neigh.source_index for neigh in list_neighbors])
+        list_source_indices.append(source_indices)
+
+        dest_indices = torch.tensor([neigh.destination_index for neigh in list_neighbors])
+        list_dest_indices.append(dest_indices)
+
+        displacements = torch.stack([torch.from_numpy(neigh.displacement) for neigh in list_neighbors])
+        list_displacements.append(displacements)
+
+        shifts = torch.stack([torch.from_numpy(neigh.shift) for neigh in list_neighbors])
+        list_shifts.append(shifts)
+
+    expected_source_indices = torch.nn.utils.rnn.pad_sequence(list_source_indices,
+                                                              batch_first=True,
+                                                              padding_value=INDEX_PADDING_VALUE)
+
+    expected_destination_indices = torch.nn.utils.rnn.pad_sequence(list_dest_indices,
+                                                                   batch_first=True,
+                                                                   padding_value=INDEX_PADDING_VALUE)
+
+    expected_displacements = torch.nn.utils.rnn.pad_sequence(list_displacements,
+                                                             batch_first=True,
+                                                             padding_value=POSITION_PADDING_VALUE)
+
+    expected_shifts = torch.nn.utils.rnn.pad_sequence(list_shifts,
+                                                      batch_first=True,
+                                                      padding_value=POSITION_PADDING_VALUE)
+
+    return expected_source_indices, expected_destination_indices, expected_displacements, expected_shifts
+
+
+# This test might be slow because KeOps needs to compile some stuff...
```

Comment:
this test takes about 25 seconds on my computer. Should we flag it as 'not_on_github'?

Reply:
This test was slow because I was extracting the permutation of indices with a slow algo. I reimplemented (and tested) a faster algorithm for this: the test is now fast.